### ISSUE TITLE
CSP violation: inline-style

### ISSUE THREAD

**Issue by AntoineAugusti** [vega/vega-embed#1214]
At least the following lines
https://github.com/vega/vega-embed/blob/bdc70735450f10acb19c2439426d272adb108ede/src/embed.ts#L311
https://github.com/vega/vega-embed/blob/bdc70735450f10acb19c2439426d272adb108ede/src/embed.ts#L401

raise an error when a CSP is set.

> embed.ts:311 Refused to apply inline style because it violates the following Content Security Policy directive: "default-src 'self' 'nonce-8IBTHwOdqNKAWeKl7plt8g=='". Either the 'unsafe-inline' keyword, a hash ('sha256-9uoGUaZm3j6W7+Fh2wfvjI8P7zXcclRw5tVUu3qKZa0='), or a nonce ('nonce-...') is required to enable inline execution. Note also that 'style-src' was not explicitly set, so 'default-src' is used as a fallback.

## Minimal reproduction

```html
<!doctype html>
<html>
  <head>
    <meta charset="utf-8" />
    <meta name="viewport" content="width=device-width, initial-scale=1" />
    <meta http-equiv="Content-Security-Policy" content="default-src 'self' 'nonce-8IBTHwOdqNKAWeKl7plt8g=='" />
    <title>Vega-Embed for Vega-Lite</title>

    <script nonce="8IBTHwOdqNKAWeKl7plt8g==" src="https://cdn.jsdelivr.net/npm/vega@5/build/vega.min.js"></script>
    <script nonce="8IBTHwOdqNKAWeKl7plt8g==" src="https://cdn.jsdelivr.net/npm/vega-lite@5/build/vega-lite.min.js"></script>
    <script nonce="8IBTHwOdqNKAWeKl7plt8g==" src="https://cdn.jsdelivr.net/npm/vega-embed@6/build/vega-embed.min.js"></script>
  </head>

  <body>
    <div id="vis"></div>
    <script nonce="8IBTHwOdqNKAWeKl7plt8g==">
      async function run() {
        const spec = {
          $schema: "https://vega.github.io/schema/vega-lite/v5.json",
          data: {
            values: [
              { a: "A", b: 28 },
              { a: "B", b: 55 },
              { a: "C", b: 43 },
              { a: "D", b: 91 },
              { a: "E", b: 81 },
              { a: "F", b: 53 },
              { a: "G", b: 19 },
              { a: "H", b: 87 },
              { a: "I", b: 52 },
            ],
          },
          mark: "bar",
          encoding: {
            x: { field: "a", type: "nominal" },
            y: { field: "b", type: "quantitative" },
          },
        };

        const result = await vegaEmbed("#vis", spec, {
          ast: true,
        });
      }
      run();
    </script>
  </body>
</html>

```

**Comment by domoritz** [comment]
Thanks for the issue. Can you send a pull request to improve the logic?

**Comment by AntoineAugusti** [comment]
@domoritz I'm not sure I can. I don't write much JS/TS on a daily basis.

The issue comes from the `style#vega-tooltip-style` and `style#vega-embed-style`. I see two options.

## Updating the CSP

Users could specify sha values for these `<style>` elements and they would load but these SHA values will change if the default style changes.

Changing the CSP in my example to this does the trick
```html
<meta http-equiv="Content-Security-Policy" content="default-src 'self' 'nonce-8IBTHwOdqNKAWeKl7plt8g==' 'sha256-9uoGUaZm3j6W7+Fh2wfvjI8P7zXcclRw5tVUu3qKZa0=' 'sha256-MmUum7+PiN7Rz79EUMm0OmUFWjCx6NZ97rdjoIbTnAg='" />
```

## Passing a nonce

We could add a new option to vega-embed, `nonce` and users would provide a value. `vega-embed` and `vega-tooltip` would set the nonce attribute to this value when creating style elements.

https://github.com/vega/vega-embed/blob/bdc70735450f10acb19c2439426d272adb108ede/src/embed.ts#L305
https://github.com/vega/vega-tooltip/blob/fc93da1bf7d77817230e6d2f41812e6d4506616a/src/Handler.ts#L40

What do you think?

**Comment by domoritz** [comment]
Or would it be easier to just ask people to set the style themselves (and have an option to disable styles added from embed)? This is the first time this issue has come up so I am a bit hesitant to add more complex configurations to embed and tooltips.

**Comment by AntoineAugusti** [comment]
Right, that would be another option! We went with the SHA fix for now on our side

**Comment by vdonato** [comment]
Wow the timing on this issue being filed is a bit uncanny 😆

Hi @domoritz! It turns out we're also running into this issue while trying to make some changes to the Streamlit web client to be more CSP-friendly (styles tags injected by `vega-embed` and `vega-tooltip` are the last obstacles to this). It seems that it's already possible to disable injecting these inline styles via options that can be set when calling the `embed` function, and we'd slightly prefer this approach over setting nonces in inlined styles since we've been wanting more fine grained control over this styling anyway.

The only thing that's necessary to confirm before doing this is that the classnames / IDs in [vega-embed.scss](https://github.com/vega/vega-embed/blob/main/vega-embed.scss) and [vega-tooltip.scss](https://github.com/vega/vega-tooltip/blob/main/vega-tooltip.scss) will be stable going forward. Looking at the git blames for both, it seems like a safe assumption, but wanted to double check.

**Comment by domoritz** [comment]
I don't think I want to make guarantees for the class names for forever but I don't have any intentions of changing them without strong reasons so you can assume they don't change. I think changing the style in any major way should be a major version bump at this point.

**Comment by vdonato** [comment]
👍 thanks -- things being reasonably stable and the major version being bumped for any significant changes should be more than sufficient for us.

**Comment by domoritz** [comment]
Is this sufficiently addressed by #1242 for now?

**Comment by vdonato** [comment]
I think so.

There may be some reasons why other projects would want avoid duplicating styles and have vega continue to inline them, but moving the styles out + the changes in #1242 are sufficient for the Streamlit client, at least.
